Orphans in split breaks preview button in Studio

awaisdar001 opened the pull request:
[PLAT-833](https://openedx.atlassian.net/browse/PLAT-833)

### Reproduce
1. Go to unit: [MITx test course](https://studio.edge.edx.org/container/block-v1:MITx+S101+2015_T1+type@vertical+block@e9ec465f8bc949ae930ad5d8047e6e81)
2. Press Preview 

Expected behavior: would see content 
Actual Behavior: page not found 

To create in a fresh course: 

1. Create a new course and add a Problem competent.
2. Export course 
3. Change chapter file name in chapter folder. example. (0c1ec994774145e1a11d82d108e68203.xml to changed.xml) 
4. Change course.xml reference ex. (url_name="0c1ec994774145e1a11d82d108e68203" to url_name="changed") 
5. Import course 
6. Go to Studio and click on preview 

LMS is fine, and you can navigate to the unit on preview.edx.org.

awaisdar001 commented:
@adampalay @cpennington @nasthagiri Please review.

adampalay commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -1134,8 +1134,19 @@ def get_parent_location(self, locator, **kwargs):
+                BlockKey(id=c.block_id, type=c.block_type)
+                for c in self.get_orphans(locator.course_key)
+            ]
+            if orphan_block_keys:
```
If both parents are orphans, do we want to return `None`? Or are we not really worried about this case?

awaisdar001 replied:
If both parents are non then `parents_ids` would become and empty list, and the following instruction will return `None`
```
if len(parent_ids) == 0:
    return None
```

adampalay commented on cms/djangoapps/contentstore/tests/test_orphan.py:
```diff
@@ -128,3 +131,50 @@ def test_not_permitted(self, default_store):
+
+    def test_orphan_parents(self):
+        store = self.store
+        with store.default_store(ModuleStoreEnum.Type.split):
```
If I'm understanding this correctly, you want a course tree that looks like:
```
 course
    |
 chapter
   |
 seq     seq
    \    /
     html
```

And you want to make sure that calling `get_parent` on `html` never returns the orphan parent? We actually have such an element already in `TestOrphanBase`: https://github.com/edx/edx-platform/blob/master/cms/djangoapps/contentstore/tests/test_orphan.py#L48-L52.

Just use that one

awaisdar001 replied:
Copy, Now using the the same element.

adampalay commented:
I'm worried about this scenario:
```
 course
    |
chapter  chapter
   |      |
 seq     seq
    \    /
     html
```
Neither parent of `html` is strictly an orphan.

awaisdar001 commented:
@adampalay  I have wrote an other test to verify the scenario you have mentioned. What I have done is, If there are multiple parents of an component, first we filter the orphans parents out. After filtering the orphans, if there are still multiple parents, then we can get, verify and filter grand parents which are orphans.

adampalay commented:
I'm not crazy about this solution, since it's also possible that a module's grandparent isn't an orphan, but its great-grandparent is. You may have to write something recursive that finds all of a module's terminal grandparents, and selects the parent that's descendent from the course root.

Like make a list of paths up from the module, remove the ones that don't end up at the course root, and choose a parent from one of the remaining paths.

awaisdar001 commented:
@adampalay  Interesting; Let me think how we can have this; Let me also look at the `path_to_location` method in detail.

adampalay commented:
I don't think `path_to_location` is going to work—it looks like it assumes that each module has one parent.

awaisdar001 commented:
@adampalay `path_to_location` internally calls `get_parent_location`, which means that this method also depends upon `get_parent_location` & if an orphan is selected (which was an issue) this method raises an exception of `NoPathToItem`.

awaisdar001 commented:
I have changed the logic of getting the `course_orphan` & filtering orphan parents & grand-parents to a recursive method. This method verifies that if an xblock has a path to the root or not. It return `boolean` whether or not a xblock has path to the course root.

adampalay commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -1128,14 +1128,36 @@ def get_parent_location(self, locator, **kwargs):
 
         :param locator: BlockUsageLocator restricting search scope
         """
+        def is_root_block(xblock, is_root=False):
```
nit: I wouldn't put `is_root` as a kwarg

adampalay replied:
This function should really be called `has_path_to_root`.

I like the idea here, that `get_parent_location` will only return parents in the course tree. This may help prevent many other orphan-related bugs! However, I also worry about if orphan's children ever need this information? I can't think of a scenario off the top of my head where they would need it.

@cpennington , I know you wanted to see this work happening in [`find_path_to_course`](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/search.py#L41). What do you think of instead putting it here?

cpennington replied:
My concern would be around performance. If walking to the root is fast, then I like the idea of filtering parents to ignore the ones in orphaned chains.

cpennington replied:
The other question is: is this still going to be broken in old-mongo courses, and do we care?

adampalay replied:
For the first question, how can we measure this? I'd imagine this is fairly fast, since we're not making any database calls.

For the second, not as much? Even if we do want to fix it, fixing it first for new mongo would make a difference for the courses that are actually currently running and are currently under development.

cpennington replied:
Oh, good point. Sorry, I was thinking this was the general case, not split only. Walking up to the root in split is fast enough.

adampalay commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -1128,14 +1128,36 @@ def get_parent_location(self, locator, **kwargs):
+                return True
+
+            for xblock_parent in xblock_parents:
+                is_root = is_root_block(xblock_parent, is_root)
```
This would just be simpler to write as
```
return any(is_root_block(xblock_parent) for xblock_parent in xblock_parents)
```
Then you avoid unnecessary iterations if it turns out that one of the parents has a path to the root.

adampalay commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -1128,14 +1128,36 @@ def get_parent_location(self, locator, **kwargs):
         """
+        def is_root_block(xblock, is_root=False):
+            """ Check recursively if an xblock has a path to the course root """
+            xblock_parents = self._get_parents_from_structure(xblock, course.structure)
```
we might be able to avoid confusion down the line if we pass `course` as an argument to this function

awaisdar001 commented:
I have updated the PR with `mongo` implementation as well. Mongo uses cache, so I have added a method to check if the selected parent is valid or not (going to the top of the course root). If any orphan found do not fetch the parent from they cache but check it from the `collection`. Any xblock is not an orphan If parents are `None` and `block_type` is in `['course', 'library']`.

There still seems some changes and modification left which I would like to discuss.

@adampalay I have addressed your comments.

awaisdar001 commented:
jenkins run js

awaisdar001 commented:
jenkins run lettuce

adampalay commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -1128,14 +1128,34 @@ def get_parent_location(self, locator, **kwargs):
+            """ Check recursively if an xblock has a path to the course root """
+            xblock_parents = self._get_parents_from_structure(xblock, course.structure)
+
+            if len(xblock_parents) == 0 and xblock.type in ["course", "library"]:
```
@cpennington  is there a "root class" or anything like that we can check against, instead of hardcoding this here?

adampalay commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -1128,14 +1128,34 @@ def get_parent_location(self, locator, **kwargs):
 
         :param locator: BlockUsageLocator restricting search scope
         """
+        def has_path_to_root(xblock, course):
```
@awaisdar001 , I think this method could be useful in a plethora of places. For example, it could be nice to use this to decide if a discussion module is in the course tree.

adampalay commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -1554,8 +1554,23 @@ def _get_raw_parent_location(self, location, revision=ModuleStoreEnum.RevisionOp
 
         parent_cache = self._get_parent_cache(self.get_branch_setting())
+
+        def has_path_to_root(xblock):
```
Is this expensive? I'm less worried about fixing this in old mongo, but if this isn't expensive to implement, great :)

adampalay commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -1607,7 +1622,8 @@ def cache_and_return(parent_loc):  # pylint:disable=missing-docstring
             # since we sorted by SORT_REVISION_FAVOR_DRAFT, the 0'th parent is the one we want
-            if published_parents > 1:
+            # If multiple draft parents found, select the non-orphaned
+            if published_parents > 1 or all_parents > 1:
```
See comment above:
```python
# there could be 2 different parents if
#   (1) the draft item was moved or
#   (2) the parent itself has 2 versions: DRAFT and PUBLISHED
#  if there are multiple parents with version PUBLISHED then choose from non-orphan parents
```
We shouldn't be concerned about `all_parents` here, I don't think.
Also `all_parents` is a list, while `published_parents` is a number. Also, it looks like orphaned parents get filtered out in the next line.

adampalay commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -1554,8 +1554,23 @@ def _get_raw_parent_location(self, location, revision=ModuleStoreEnum.RevisionOp
+            """ Check recursively the cache to get the parent & findout if xblock has a path
+            to the course root """
+
+            xblock_parent = parent_cache.get(unicode(xblock))
```
does the `parent_cache` assume one parent per item?

awaisdar001 replied:
@adampalay Yes it return only one item.

adampalay commented:
@awaisdar001 , I don't think we need to fix this in old mongo as a part of this PR; let's save it for another, lower-priority ticket. For now it's much more important to patch this for split mongo. Also, if you're able to add `has_path_to_root` as a more generalizable modulestore method, we could use it when [fetching all discussion modules for a course](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/django_comment_client/utils.py#L101) and so avoid the orphan discussion module issue too.

awaisdar001 commented:
@adampalay Sounds fair enough. With this change in the mongo; I was still figgring out how to fix the other tests that were failing in Mongo. So, Lets have the fix and the split and and get a code review.

awaisdar001 commented:
@mushtaqak I changed your test because it seems to me the `sequential` is an orphan. Please suggest if this is a right change?

awaisdar001 commented:
@adampalay  I moved  `has_path_to_root` method to be the `SplitMongoModuleStore` class.

adampalay commented:
@awaisdar001 , this is awesome!

:+1:

adampalay commented on cms/djangoapps/contentstore/tests/test_orphan.py:
```diff
@@ -128,3 +131,88 @@ def test_not_permitted(self, default_store):
+
+        # Get path of the `multi_parent_html` & Verify it should not be None
+        path = path_to_location(self.store, multi_parent_html.location)
+        self.assertIsNotNone(path)
```
should we assert that len(path) == 1?

cpennington replied:
I would be tempted to assert on the format of the `path_to_location` output.

adampalay commented on cms/djangoapps/contentstore/tests/test_orphan.py:
```diff
@@ -128,3 +131,88 @@ def test_not_permitted(self, default_store):
+
+        path = path_to_location(self.store, html.location)
+        self.assertIsNotNone(path)
+        self.assertEquals(path[0], course.id)
```
should we assert len(path) == 1?

awaisdar001 replied:
The reason I think the `len(path)` should not be `==1` because `path_to_location` method returns a tuple. 
> a tuple (course_id, chapter, section, position) is returned which is suitable for the courseware index view.

If at any point, `get_parent_location` find outs that there is an orphan in xblocks' parent/super parent it will return `None`. So its either a tuple or None which is returned.
[common/lib/xmodule/xmodule/modulestore/search.py#L113](https://github.com/edx/edx-platform/blob/756008aa9bd03ecd147ae98271a16fb640d59945/common/lib/xmodule/xmodule/modulestore/search.py#L113)

adampalay commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -1120,6 +1120,15 @@ def _block_matches_all(block_data):
         else:
             return []
 
+    def has_path_to_root(self, xblock, course):
```
nit: `xblock` maybe should be `block_key`?

awaisdar001 replied:
Added an assert `assert isinstance(xblock, BlockKey)` although we can would like to raise `ItemNotFoundError`, we can also use the following.

```
if not isinstance(xblock, BlockKey):
    raise ItemNotFoundError(xblock)
```

adampalay replied:
I don't see why we need this assert; I just thought `block_key` was a more descriptive name than `xblock`

awaisdar001 replied:
Oh i see.

awaisdar001 replied:
Comment addressed.

mushtaqak commented:
@awaisdar001 this should be okay. 

It was fair that the test was failing as `create_item` created the orphan sequential, so with your change to check `xblock` that has root path,  `publish` method was not called while deleting the unit in the unit test. 

Nice catch on `create_child` :)

adampalay commented:
:+1:

awaisdar001 commented:
@cpennington @nasthagiri a review please?

cpennington commented:
:+1:, once the assertions are more specific than just checking for `is not None`.

awaisdar001 commented:
@adampalay A last look? I have addressed cale's comments. Should we proceed to merge this?

adampalay commented:
:+1: , this is great :)
